backend/s3/s3hash/s3hash.go

```diff
@@ -0,0 +1,140 @@
+// Package s3hash implements the AWS muliparted hash, which is a modified MD5.
+// See https://stackoverflow.com/questions/12186993/what-is-the-algorithm-to-compute-the-amazon-s3-etag-for-a-file-larger-than-5gb
+//
+// Say you uploaded a 14MB file to a bucket without server-side encryption, and your part size is 5MB.
+// Calculate 3 MD5 checksums corresponding to each part, i.e. the checksum of the first 5MB,
+// the second 5MB, and the last 4MB. Then take the checksum of their concatenation.
+// MD5 checksums are often printed as hex representations of binary data, so make sure you take the MD5
+// of the decoded binary concatenation, not of the ASCII or UTF-8 encoded concatenation.
+// When that's done, add a hyphen and the number of parts to get the ETag.
+//
+// For example, a multipart Etag can be built with the code below.
+// 	    hs := s3hash.New(8 * Mi)
+// 	    for {
+// 	        hs.Write(data)
+//      }
+//      Etag := hex.EncodeToString(hs.Sum(nil)) + "-" + strconv.Itoa(hs.GetPartsCount())
+//      // result be like — 60b725f10c9c85c70d97880dfe8191b3-33
+package s3hash
+
+import (
+	"crypto/md5"
+	"encoding"
+	"hash"
+)
+
+type S3Hash struct {
```

Comment:
What do you think? Here we need a lock in case of use with goroutines?

Reply:
No locks needed - hashes aren't expected to be thread-safe